File: C++/reverse_words_in_a_file/reverse_words_in_a_file.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <algorithm>

// Function to reverse individual words while keeping spaces/structure intact
std::string reverseWordsInLine(const std::string& line) {
    std::string result = "";
    std::string word = "";

    for (char ch : line) {
        // If we encounter a space or punctuation that splits words,
        // reverse the accumulated word and add it to the result
        if (std::isspace(ch)) {
            std::reverse(word.begin(), word.end());
            result += word;
            result += ch; // Keep the original whitespace
            word = "";    // Reset for the next word
        } else {
            word += ch;
        }
    }

    // Don't forget the last word if the line didn't end with a space
    if (!word.empty()) {
        std::reverse(word.begin(), word.end());
        result += word;
    }

    return result;
}

int main() {
    std::string inputFilename = "harry_potter.txt";
    std::string outputFilename = "output.txt";

    std::ifstream inputFile(inputFilename);
    std::ofstream outputFile(outputFilename);

    // Check if the input file exists and can be opened
    if (!inputFile.is_open()) {
        std::cerr << "Error: Could not open the input file '" << inputFilename << "'." << std::endl;
        return 1;
    }

    // Check if the output file can be created/opened
    if (!outputFile.is_open()) {
        std::cerr << "Error: Could not create or open the output file '" << outputFilename << "'." << std::endl;
        inputFile.close();
        return 1;
    }

    std::string line;
    // Read the input file line by line
    while (std::getline(inputFile, line)) {
        std::string processedLine = reverseWordsInLine(line);
        outputFile << processedLine << "\n";
    }

    std::cout << "File processing completed successfully!" << std::endl;
    std::cout << "Transformed text saved to '" << outputFilename << "'." << std::endl;

    // Close the file streams
    inputFile.close();
    outputFile.close();

    return 0;
}

File: python/reverse_words_in_a_file/reverse_words_in_a_file.py
def reverse_words_in_file(input_file, output_file):
    # Open the source file for reading and the new file for writing
    with open(input_file, 'r', encoding='utf-8') as infile, \
         open(output_file, 'w', encoding='utf-8') as outfile:
        
        # Process the file line by line to keep it memory-efficient
        for line in infile:
            # Strip the trailing newline character, if it exists
            stripped_line = line.rstrip('\n')
            
            # Split the line into individual words
            words = stripped_line.split(' ')
            
            # Reverse the letters of each word
            reversed_words = [word[::-1] for word in words]
            
            # Join the reversed words back into a sentence
            new_line = ' '.join(reversed_words)
            
            # Write the modified line with its original newline character back
            outfile.write(new_line + '\n')

input_file = "harry_potter.txt" # This is the file to be read
output_file = "output.txt"      # This is the output file
reverse_words_in_file(input_file, output_file)

